api/task_manager.cc

```diff
@@ -0,0 +1,130 @@
+    res.progress_completed = status.task_status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        return make_exception_future<json::json_return_type>(httpd::bad_param_exception("Invalid task id"));
+    }
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module();
+    return task->get_progress().then([s] (task_manager::task::progress progress) mutable {
+        s.task_status.progress.completed = progress.completed;
+        s.task_status.progress.total = progress.total;
+        return make_ready_future<json::json_return_type>(make_status(s));
+    });
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return make_ready_future<json::json_return_type>(ctx.tm.local().get_modules());
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return do_with(req.get(), [&ctx] (auto &req) {
+            return ctx.tm.map_reduce0([&req] (task_manager& tm) {
+                std::vector<tm::task_stats> res{};
+                std::string m = req->param["module"];
+                if (tm.get_tasks_per_module().contains(m)) {
+                    for (auto& [task_id, task] : tm.get_tasks_per_module()[m]) {
+                        if (filter_tasks(task, req->query_parameters)) {
+                            tm::task_stats ts;
+                            ts.task_id = task->get_status().id.to_sstring();
+                            ts.state = task->get_status().state;
+                            res.push_back(ts);
+                        }
+                    }
+                }
+                return res;
+            }, std::vector<tm::task_stats>(), concat<tm::task_stats>);
+        }).then([] (const std::vector<tm::task_stats>& res) {
+            return make_ready_future<json::json_return_type>(res);
+        });
+    });
+
+    tm::get_task_status.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return ctx.tm.local().retrieve_task(id).then(retrieve_status);
+    });
+
+    tm::abort_task.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return ctx.tm.invoke_on_all([id] (task_manager& tm) {
+            auto task = tm.get_task_this_shard(id);
+            if (task) {
+                task->abort();
+            }
+        }).then([]{
+             return make_ready_future<json::json_return_type>(json_void());
+        });
+    });
+
+    tm::wait_task.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return ctx.tm.local().wait_task(id).then(retrieve_status);
```

Comment:
I'd rather open code `task_manager::wait_task` here.
lookup_task_on_all_shards can be used to find the task, then we can `co_await task->done()` and finally get the final status and unregister the task.

One thing we need to be careful about is controlling exactly when `done()` returns an error, if at all.
If an error happened during the execution of the task, it should be represented in the task status,
but the `done()` future should succeed, otherwise we'll fail here and return an error to the api call, which is not what we want to do.

A rest_api error should happen if something wrong went in the rest_api function (like failing to find the task_id), not in the task itself.